Fixes importer issue which could hang shutdown during panic

tinwelint opened the pull request:
this would prevent import tool from shutting down, but more importantly
it would prevent the error to be shown.

This fix treats a call to close() before all processing is done in
ParallelInputEntityDeserializer as a panic and delegates that flag to
its executor so that all processors can now abort their processing
accordingly.

tinwelint committed:
Fixes importer issue which could hang shutdown during panic

this would prevent import tool from shutting down, but more importantly
it would prevent the error to be shown.

This fix treats a call to close() before all processing is done in
ParallelInputEntityDeserializer as a panic and delegates that flag to
its executor so that all processors can now abort their processing
accordingly.

MishaDemianenko commented on community/import-tool/src/test/java/org/neo4j/tooling/EntityDataGenerator.java:
```diff
@@ -84,7 +84,7 @@ public long position()
     {
         super.close();
-        processing.shutdown( false );
+        processing.shutdown( 0 );
```
Would be really nice if we will have plain `shutdown()` in case everything is fine and `shutdown(TaskPolicy taskPolicy)` that take enum value as param instead of int.
Then in simple case you do not need even think what the number you need to pass that and in case you actually need to pass something - you can see enum values instead of trying yo find int value

burqen replied:
+1 on this

tinwelint committed:
Propagates slurp failures to user of processing

so that failures doesn't just hang the processing, but not providing
more jobs and doesn't signal that it has failed.

burqen reviewed (changes requested):
This looks good. Would be cleaner with the suggestion from @MishaDemianenko

tinwelint committed:
Simplified TaskExecutor/TicketedProcessing shutdown

by removing shutdown flags and having an explicit panic(Throwable)
method besides shutdown(). TicktedProcessing separates shutdown(flags)
into shutdown() and endOfSubmissions() instead of conflating the two
and using the previously available flags to work around and distinguish
the two cases.

This should have no observable functional difference looking at the
current usages of TaskExecutor/TicketedProcessing, only that the code
related to shutdown should be easier to reason about.

tinwelint commented:
@burqen Simplified shutdown now, feels much better (I was never happy about the shutdown flags)

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/executor/TaskExecutor.java:
```diff
@@ -46,23 +43,31 @@
 
     /**
-     * @return {@code true} if {@link #shutdown(int)} has been called, otherwise {@code false}.
+     * @return {@code true} if {@link #shutdown()} has been called, otherwise {@code false}.
```
From  this comment it is not clear to be if `isShutdown` will return true or false if `panic` has been called. Can we please clarify this?

burqen commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/executor/DynamicTaskExecutor.java:
```diff
@@ -164,24 +160,35 @@ private void notifyProcessors()
+    @Override
+    public void panic( Throwable panic )
+    {
+        this.panic = panic;
```
Multiple calls to `panic` will overwrite each other. That means original cause of failure will be overwritten by subsequent failures.

We could chain them together, although we do not _know_ that they are related. Another idea would be to keep a list of panics and report them all in `assertHealthy`.

chrisvest replied:
… or add subsequent panics as suppressed exceptions to the original.

burqen reviewed (changes requested):
Second review done. Please have a look.

tinwelint committed:
Can capture multiple panics in DynamicTaskExecutor

by adding them as suppressed to the first one.

burqen commented:
@tinwelint Looking good. Merge once green!

burqen commented:
@tinwelint Had to revert this merge because of test failure in `ParallelInputEntityDeserializerTest.shouldTreatExternalCloseAsPanic`.
Branch should be rebased on 3.0, fixed and put into a new PR.
